sycl/doc/extensions/experimental/sycl_ext_oneapi_graph_fusion.asciidoc

```diff
@@ -0,0 +1,636 @@
+or allocated device memory is accessed by no more than one work-item in the
+kernel submitted by this command-group (in case the property is specified on an
+accessor) or in any kernel in the graph (in case the property is specified on a
+buffer or an USM pointer). Implementations may treat this as a hint to promote
+the elements of the buffer or allocated device memory to private memory (see below
+section on local and private internalization).
+
+The application also asserts that the updates made to the buffer or allocated
+device memory by the kernel submitted by this command-group (in case the
+property is specified on an accessor) or in any kernel in the graph (in case the
+property is specified on a buffer or an USM pointer) may not be available for
+use after the fused kernel completes execution. Implementations may treat this
+as a hint to not initialize or write back the final result to global memory.
+
+If different properties (local or private) are applied to accessors to the same
+buffer, the resolution rules specified below apply. The property is not
+prescriptive, implementations are free to not perform internalization and it is
+no error if they do not perform internalization. Implementations can provide a
+diagnostic message in case internalization was not performed through an
+implementation-specified mechanism, but are not required to do so.
+
+==== Device information descriptors
+
+To support querying whether a SYCL device and the underlying platform support
+kernel fusion for graphs, the following device information descriptor is added
+as part of this extension proposal. 
+
+```c++
+sycl::ext::oneapi::experimental::info::device::supports_fusion
+```
+
+When passed to `device::get_info<...>()`, the function returns `true` if the
+SYCL `device` and the underlying `platform` support kernel fusion for graphs.
+
+
+=== Linearization
+
+In order to be able to perform kernel fusion, the commands in a graph must be
+arranged in a valid sequential order. 
+
+A valid _linearization_ of the graph is an order of the commands in the graph
+such that each command in the linearization depends only on commands that appear
+in the sequence before the command itself. 
+
+The exact linearization of the dependency DAG (which generally only implies a
+partial order) is implementation defined. The linearization should be
+deterministic, i.e. it should yield the same sequence when presented with the
+same DAG.
+
+=== Synchronization in kernels
+
+Group barriers semantics do not change in the fused kernel and barriers already
+in the unfused kernels are preserved in the fused kernel. Despite this, it is
+worth noting that, in order to introduce synchronization between work items in a
+same work-group executing a fused kernel, a barrier is added between each of the
+kernels being fused. This automatic insertion of additional barriers can be
+deactivated through the property defined above.
+
+=== Limitations
```

Comment:
Do we need to restrict the use of this graphs entry-point which updates an executable graph with the inputs/outputs of another topologically identical graph?
```
void command_graph<graph_state::executable> update(const command_graph<graph_state::modifiable>& graph)
````
Guessing if an executable graph has already been fusued you won't be able to swap out an input with another using different internalization properties.

There's also a pending PR - https://github.com/reble/llvm/pull/77 - that proposes another entry-point with a more direct update.

Reply:
Interesting point, thanks for bringing this up. 

The question here (next to some technical/implementation details) probably is, how computationally expensive the `update` operation is allowed to be. The graph proposal does not mention computational complexity or whether the update is finished when returning from `update`. 

If `update` is allowed to be a computationally expensive operation, updating a fused graph with new inputs with different internalization properties could simply trigger re-compilation (=fusion) of the graph, yielding a new fused graph. In that case, no additional restrictions would need to be applied to the `update` API. 

In the current implementation of kernel fusion, we also use a cache to avoid re-compilation in case the same sequence of kernels is fused repeatedly and the internalization properties (and some other context) have not changed. The update scenario would be somewhat similar, only triggering re-compilation on update if relevant parameters, including internalization properties, have changed.

Reply:
I envisage that the computational complexity ultimately will depend on backend support, e.g [cl_khr_command_buffer_mutable_dispatch](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_command_buffer_mutable_dispatch), Vulkan, or CUDA Graph backends lets you update the arguments to an existing command-buffer easily, but Level Zero or vanilla OpenCL backends do not provide a mechanism so the command-lists would have to be recreated in a slower path.

The current PI interface also doesn't provide a way to expose this, so our extensions to PI for graphs will need to expose kernel arguments differently so that cl_khr_command_buffer_mutable_dispatch/Vulkan/CUDA-Graphs could take advantage it.

So the `update()` entry-point is definitely allowed to be expensive, equivalent to finalizing again. In the immediate future of our graphs implementation it will be the slower emulated path until we get round to adding the relevant PI, which isn't of my top of our priority list as it's a change for performance rather than correctness.

Reply:
> So the update() entry-point is definitely allowed to be expensive, equivalent to finalizing again

In that case, re-compiling on an update with different internalization properties (or otherwise different context) seems like the best route to go, which means that no additional restrictions would need to be applied to the `update` API.